runtime/doc/lsp.txt

```diff
@@ -0,0 +1,306 @@
+
+                                                          *lsp#request_async()*
+lsp#request_async({method} [, {arguments} [, {callback} [, {filetype}]]])
+    Creates and sends a request to the server for [filetype]. This request
+    will not block and will call desired the [callback] upon completion.
+
+    Returns the request ID.
+
+    See |lsp#request()| for more information.
+
+    Example:
+
+autocmd CursorHold *
+    \ :lua vim.lsp.plugin.request_async(
+    \     'textDocument/hover',
+    \     vim.lsp.util.get_text_document_params(),
+    \     vim.lsp.util.get_filetype()
+    \ )
+
+================================================================================
+                                                         *lsp-built-in-callback*
+
+The following requests have built-in callbacks defined to handle the response
+in an idiomatic vim fashion.
+
+  textDocument/definition~
+  textDocument/hover~
+  textDocument/publishDiagnostics~
+  textDocument/references~
+
+vim.lsp.config.callbacks.set_all_builtin_callbacks()
+  You can set all built-in callbaks
+
+vim.lsp.config.callbacks.set_builtin_callback({method})
+    You can set a specify built-in callback.
+        Parameters:~
+            {method}: The name of the request, as specified by the
+            language server protocol.
+                Example: 'textDocument/hover'
+
+vim.lsp.config.callbacks.set_builtin_callbacks({methods})
+    You can set specify built-in callbacks.
+        Parameters:~
+            {method}: The name of the request, as specified by the
+            language server protocol.
+                Example: {'textDocument/hover', 'textDocument/publishDiagnostics'}
+
+
+================================================================================
+                                                         *lsp-common-behavior*
+
+Many of the Language Server Protocol messages are similar to common commandds
+in nvim. Here are some example configurations that could integrate LSP into
+your daily workflow.
+
+  textDocument/definition~
+    This is essentially "go-to definition". Since this is generally done
+    through tags, the following mapping might be considered idiomatic.
+
+    nnoremap <buffer> <C-]> :call lsp#request_async('textDocument/definition')
+
+================================================================================
+                                                            *lsp-configuration*
+
+
+                                                              *lsp-add_check()*
+require('lsp.checks').add_check({request}, {cb})
+    Configure when methods will be requested.
+    Can be used to disable certain methods, if you want the functionality to
+    be disabled in the plugin.
+
+    Parameters:~
+        {request}: The name of the request, i.e. 'textDocument/hover'
+
+        {cb}: A callback to execute to determine if the {request} should be
+        sent.
+
+        If `false`, the {request} will never be sent.
```

Comment:
```suggestion
        If `false`, the {request} will not be sent.
```

How often / when is this called?  (I assumed it would be done before every (potential) request)

Also could/should document `client` and `request` args for the callback.